crates/rspack_loader_swc/src/lib.rs

```diff
@@ -0,0 +1,230 @@
+use swc_core::base::config::{
+  Config, ErrorConfig, FileMatcher, InputSourceMap, IsModule, JscConfig, ModuleConfig, Options,
+  SourceMapsConfig, TransformConfig,
+};
+use swc_core::base::{try_with_handler, Compiler};
+use swc_core::common::comments::SingleThreadedComments;
+use swc_core::common::{FileName, FilePathMapping, GLOBALS};
+use swc_core::ecma::transforms::base::pass::noop;
+
+#[derive(Debug, Default, Deserialize)]
+#[serde(rename_all = "camelCase", default)]
+pub struct SwcLoaderJsOptions {
+  #[serde(default)]
+  pub source_maps: Option<SourceMapsConfig>,
+
+  pub source_map: Option<SourceMapsConfig>,
+  #[serde(default)]
+  pub env: Option<swc_core::ecma::preset_env::Config>,
+
+  #[serde(default)]
+  pub test: Option<FileMatcher>,
+
+  #[serde(default)]
+  pub exclude: Option<FileMatcher>,
+
+  #[serde(default)]
+  pub jsc: JscConfig,
+
+  #[serde(default)]
+  pub module: Option<ModuleConfig>,
+
+  #[serde(default)]
+  pub minify: BoolConfig<false>,
+
+  #[serde(default)]
+  pub input_source_map: Option<InputSourceMap>,
+
+  #[serde(default)]
+  pub inline_sources_content: BoolConfig<true>,
+
+  #[serde(default)]
+  pub emit_source_map_columns: BoolConfig<true>,
+
+  #[serde(default)]
+  pub error: ErrorConfig,
+
+  #[serde(default)]
+  pub is_module: Option<IsModule>,
+
+  #[serde(rename = "$schema")]
+  pub schema: Option<String>,
+}
+
+#[derive(Debug, Default, Deserialize)]
+#[serde(rename_all = "camelCase", default)]
+struct SwcLoaderOptions {
+  pub config: Option<Config>,
+  pub source_maps: Option<SourceMapsConfig>,
+}
+
+impl From<SwcLoaderJsOptions> for Options {
+  fn from(value: SwcLoaderJsOptions) -> Self {
+    let SwcLoaderJsOptions {
+      source_maps,
+      source_map,
+      env,
+      test,
+      exclude,
+      jsc,
+      module,
+      minify,
+      input_source_map,
+      inline_sources_content,
+      emit_source_map_columns,
+      error,
+      is_module,
+      schema,
+    } = value;
+    let mut source_maps: Option<SourceMapsConfig> = source_maps;
+    if source_maps.is_none() && source_map.is_some() {
+      source_maps = source_map
+    }
+    if let Some(SourceMapsConfig::Str(str)) = &source_maps {
+      if str == "inline" {
+        source_maps = Some(SourceMapsConfig::Bool(true))
+      }
+    }
+    Options {
+      config: Config {
+        env,
+        test,
+        exclude,
+        jsc,
+        module,
+        minify,
+        input_source_map,
+        source_maps,
+        inline_sources_content,
+        emit_source_map_columns,
+        error,
+        is_module,
+        schema,
+      },
+      ..Default::default()
+    }
+  }
+}
+
+#[derive(Debug)]
+pub struct SwcLoader {
+  options: Options,
+}
+
+impl SwcLoader {
+  pub fn new(options: SwcLoaderJsOptions) -> Self {
+    Self {
+      options: Options::from(options),
+    }
+  }
+}
+
+/// Get global sourcemap
+pub fn compiler() -> Arc<Compiler> {
+  static C: Lazy<Arc<Compiler>> = Lazy::new(|| {
```

Comment:
I don't think it's necessary to use the global one, as we are not supposed to share anything between each run and this would cause tangible lock contentions.

Reply:
Do you mean a new swc compiler every time
Like this?
```
  let c =  Compiler::new(Arc::from(swc_core::common::SourceMap::new(FilePathMapping::empty())));
```

Reply:
Yes

Reply:
Is there a performance penalty every time you create a new instance? I feel like there is, but I did a quick test and it doesn't seem to be😅

Reply:
I don't think there will be a performance issue here.